src/Chapter3.hs

```diff
@@ -1006,6 +1170,143 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data Action
+  = Attack
+  | DrinkPotion
+  | CastSpell
+  | RunAway deriving (Show)
+
+data Monster = Monster
+  { monsterHealth  :: Int
+  , monsterAttack  :: Int
+  , monsterActions :: [Action]
+  } deriving (Show)
+
+data Knight = Knight
+  { knightHealth  :: Int
+  , knightAttack  :: Int
+  , knightDefense :: Int
+  , knightActions :: [Action]
+  , knightSpell   :: Int
+  , knightPotion  :: Int
+  } deriving (Show)
+
+class Fighter a where
+  takeDamage :: a -> Int -> a
+  heal       :: a -> Int -> a
+  applySpell :: a -> Int -> a
+  isDead     :: a -> Bool
+  attack     :: (Fighter b) => a -> b -> b
+  doAction   :: (Fighter b) => a -> b -> Action -> (a, b)
+  doActions  :: (Fighter b) => a -> b -> (a, b)
+
+instance Fighter Monster where
+  takeDamage :: Monster -> Int -> Monster
+  takeDamage monster amount = monster { monsterHealth = monsterHealth monster - amount }
+
+  -- Monsters can't do this
+  heal :: Monster -> Int -> Monster
+  heal monster _ = monster
+
+  -- Monsters can't do this
+  applySpell :: Monster -> Int -> Monster
+  applySpell monster _ = monster
+
+  isDead :: Monster -> Bool
+  isDead = (<= 0) . monsterHealth
+
+  attack :: (Fighter b) => Monster -> b -> b
+  attack monster opponent = takeDamage opponent $ monsterAttack monster
+
+  doAction :: (Fighter b) => Monster -> b -> Action -> (Monster, b)
+  doAction monster opponent RunAway =
+    ( monster
+    , opponent
+    )
+  doAction monster opponent Attack =
+    ( monster
+    , takeDamage opponent $ monsterAttack monster
+    )
```

Comment:
Also, because of how you structured the `Action` type, this function is partial. If a `Monster` by accident have a `CastSpell` action in their sequence, this function fails in runtime and it might be unpleasant to debug :disappointed: